Updated the Dockerfile for splunk-debian-9 to build in multiple stages

mikedickey opened the pull request:
*** READY FOR REVIEW AND TESTING ***

Updated the Dockerfile for splunk-debian-9 to build in multiple stages, including:

* minimal-debian-9: Currently intended only for internal or experimental purposes.
                    Size: 137 MB compressed, 434 MB unpacked

* bare-debian-9: Includes full Splunk Enterprise package, but without ansible.
                 Includes all layers of minimal-debian-9.
                 Size: 385 MB compressed, 1059 MB unpacked

* splunk-debian-9: Includes full Splunk Enterprise package, with ansible.
                   Includes all layers of bare-debian-9 and minimal-debian-9.
                   Size: 385 MB compressed, 1072 MB unpacked

splunk-debian-9 should be nearly identical to upstream, with following exceptions:

* Splunk is now completely unpackaged into the image layers and /tmp/splunk.tgz has
  been eliminated.

* The /opt/splunk/etc/ directory is still empty to accomodate volume mounts and
  "in-place upgrade" with all etc files now located in /opt/splunk-etc/

* A /sbin/updateetc.sh script is included that checks differences in splunk.version
  and pours /opt/splunk-etc over /opt/splunk/etc if not found or different

I believe this still requires updates to entrypoint.sh to trigger
/sbin/updateetc.sh and possibly ansible modifications as well.

Commit:
Updated the Dockerfile for splunk-debian-9 to build in multiple stages, including:

* minimal-debian-9: Currently intended only for internal or experimental purposes.
                    Size: 137 MB compressed, 434 MB unpacked

* bare-debian-9: Includes full Splunk Enterprise package, but without ansible.
                 Includes all layers of minimal-debian-9.
                 Size: 385 MB compressed, 1059 MB unpacked

* splunk-debian-9: Includes full Splunk Enterprise package, with ansible.
                   Includes all layers of bare-debian-9 and minimal-debian-9.
                   Size: 385 MB compressed, 1072 MB unpacked

splunk-debian-9 should be nearly identical to upstream, with following exceptions:

* Splunk is now completely unpackaged into the image layers and /tmp/splunk.tgz has
  been eliminated.

* The /opt/splunk/etc/ directory is still empty to accomodate volume mounts and
  "in-place upgrade" with all etc files now located in /opt/splunk-etc/

* A /sbin/updateetc.sh script is included that checks differences in splunk.version
  and pours /opt/splunk-etc over /opt/splunk/etc if not found or different

I believe this still requires updates to entrypoint.sh to trigger
/sbin/updateetc.sh and possibly ansible modifications as well.

arctan5x commented:
Would minimal-debian-9 have a tar for Splunk enterprise? If not, what are the differences between minimal-debian-9 and base-debian-9?

mikedickey commented:
> Would minimal-debian-9 have a tar for Splunk enterprise? If not, what are the differences between minimal-debian-9 and base-debian-9?

Currently this removes the embedded tarball entirely; everything is unpacked on the image layers at build time. Note that this would be an alternative approach to the one represented in Ryan's PR.

In testing, one thing I noticed is that only the compressed size really matters since the layers are always compressed by docker engine prior to transmission. So the act of compressing something in the image layers impacts its uncompressed size on disk but not time to download new images. I guess I'm leaning more in the camp that the benefits of reduced uncompressed size are not worth the tradeoff you pay in startup time.

Also I noticed that using a multistage build that prepares all the /opt/splunk files beforehand and uses COPY --from=package --chown=splunk:splunk seems to create more efficient image layers. Strangely even just having the chown in a separate RUN step has a material negative impact. This leads me to feel that it's better to unpack the etc files as well versus tarring them up (even without compression).

Note that "minimal" includes Splunk while "base" does not. The difference between "minimal" and "bare" is that the latter includes all files from the Splunk Enterprise tarballs, while the former only includes a subset of minimal files. This is for internal/experimental purposes -- expect lots of changes here in the future as we tune specific layers to map to minimal requirements for different roles.

mostafahussein commented:
@mikedickey If you need an extra hand I would like to help if possible. Thanks!

Commit:
Merge remote-tracking branch 'origin/develop' into multistage-build

Commit:
Updated /sbin/updateetc.sh to do nothing if /opt/splunk-etc/splunk.version doesn't exist

Added missing HEALTHCHECK and ENTRYPOINT from splunk-debian-9 image and reordered a few
things to reduce image layers

mikedickey commented:
Please note that this requires some changes to Ansible playbooks, where are contained in this PR: https://github.com/splunk/splunk-ansible/pull/116/files

Not sure best way to tell CircleCI to use that, since it's not a submodule? Guess I could temporarily update Makefile but that seems wrong..

mikedickey commented:
On my MacBook Pro, this cuts single-instance startup time in half: from > 1 minute to < 30 seconds. Of course, it comes at the cost of having a 2x exploded on-disk footprint: about 538 MB -> 1100 MB. But the compressed image size is basically identical, so image pull times are similar.

Commit:
Modified updateetc.sh script so that it uses SPLUNK_HOME env variable

Updated remaining dockerfiles so that everything unpacks the entire
tarball to the image layers rather than including a tarball that needs
to be uncompressed at runtime (Debian and Centos / Enterprise and Forwarder)

Impact on UF startup time is much less significant: only drops by about 4 seconds

nwang92 commented on splunk/debian-9/Dockerfile:
```diff
@@ -11,61 +11,98 @@
+RUN tar -C /extras/splunk --strip 1 --wildcards --files-from=/tmp/splunk-minimal-exclude.list -zxf /tmp/${SPLUNK_FILENAME}
+RUN mv /minimal/splunk/etc /minimal/splunk-etc
+RUN mv /extras/splunk/etc /extras/splunk-etc
+RUN mkdir -p /minimal/splunk/etc/apps/_splunk_config /extras/splunk/etc/apps/_splunk_config
```
Can we condense the images layers created by the consecutive RUN's here?

nwang92 replied:
nvm, just realizing this is a multi-stage build, ignore ^

nwang92 commented on splunk/centos-7/Dockerfile:
```diff
@@ -11,61 +11,64 @@
+#
+# Download and unpack Splunk Enterprise
+#
+FROM base-debian-9:latest as package
```
Should be base-centos-7 here (if we want to continue including centos as part of this repo)

nwang92 replied:
Keep forgetting this is multi-stage - I guess it shouldn't matter if this is the just an ephemeral stage?

mikedickey replied:
Yeah, I don't think it matters in this case.. just left it to be consistent across platforms.

nwang92 commented on splunk/centos-7/Dockerfile:
```diff
@@ -11,61 +11,64 @@
 
-# Setup users and download Splunk
+# Copy ansible playbooks
+COPY splunk-ansible ${SPLUNK_ANSIBLE_HOME}
```
In terms of ordering, I think I would prefer this COPY step + the one on line 51 to be after the RUN command that sets up users/directories? 

My reasoning is during development, we're more likely to update `splunk-ansible` frequently. Putting the COPY step here would force all the subsequent steps to rebuild when it's probably doesn't necessarily need to. Granted, everything in the RUN step should be quicker now that we've removed the "download Splunk" bit, but just a matter of quality of life.

Maybe ordering should be something the following?
* RUN groupadd ...
* COPY --from=package ...
* COPY [ "splunk/common-files/updateetc.sh" ...
* COPY splunk-ansible ...

mikedickey replied:
The problem I had is that some of the RUN commands assume that the copies have taken place (chmods and such). I thought it was better for readability to just keep the COPYs together and consolidate the RUNs, and it also reduces the number of layers produced vs splitting up the RUN commands.

mikedickey replied:
And like you said, the RUN step is relatively fast now so I don't think this ordering impacts dev cycle time anymore.

nwang92 commented on splunk/debian-9/Dockerfile:
```diff
@@ -11,61 +11,98 @@
+# start a container as a particular user. So we create Splunk
+# user with pre-determined UID.
+ARG UID=41812
+ARG GID=41812
```
Seems like this might be pertinent to all images, not just `minimal`?

mikedickey replied:
This will get inherited by all the debian-9 derivatives, and is a change from currently published containers. I need to do some more testing to understand what this means and if it has a negative impact on things. In particular, I'm concerned on what it means for upgrades...

I held short from also making the same changes (TZ as well) in the centos and UF variants until I better understand the impacts. But yes, probably would be good to keep this consistent.

nwang92 commented on splunk/debian-9/Dockerfile:
```diff
@@ -11,61 +11,98 @@
+ARG UID=41812
+ARG GID=41812
+    
+RUN ln -sf /usr/share/zoneinfo/${TZ} /etc/localtime \
```
Is this a specific requirement for internal use? Or would this be more apt to build into our `base-x-x` images?

mikedickey replied:
Similar to UID/GID, I'm kind of split on these changes. Could be good to standardize on these things across the board, but also easy to change them in derived containers, if desired..

nwang92 commented:
For clarity, even with these Dockerfile changes in this repo, the only thing we publish publicly will still be the Splunk w/ Ansible playbooks?

nwang92 commented:
Also re: CircleCI, given the magnitude of the changes here, it might be best to update the default `SPLUNK_ANSIBLE_BRANCH` var in the Makefile and so the tests can run with both changes. Not a great practice, but I think it's necessary here to guarantee feature parity.

We can always revert the Makefile change in a new PR?

mikedickey commented:
> For clarity, even with these Dockerfile changes in this repo, the only thing we publish publicly will still be the Splunk w/ Ansible playbooks?

Yes, the only images that we build, push to registries and support would still be the full ones including the ansible playbooks. The others would just be available for those interested in building their own custom images, who are willing to write and provide their own scripts etc. to support custom configs and orchestration. And of course, who are willing and able to "support" that themselves.

mikedickey commented:
> Also re: CircleCI, given the magnitude of the changes here, it might be best to update the default `SPLUNK_ANSIBLE_BRANCH` var in the Makefile and so the tests can run with both changes. Not a great practice, but I think it's necessary here to guarantee feature parity.
> 
> We can always revert the Makefile change in a new PR?

Sounds good. I can make that change along with fixing any (expected) conflicts after we merge SkyKnight's "update logic" PR.

Commit:
Merge remote-tracking branch 'origin/develop' into multistage-build

Commit:
Temporary hack to use multistage-build branch of splunk-ansible (for CI)

Commit:
Updated Makefile so that it prints out the version of splunk-ansible it is using

Added a few dummy apps into the containers so that we can track docker usage via telemetry

Commit:
Fix for missing ansible version.txt

Commit:
Merge remote-tracking branch 'origin/develop' into multistage-build

Commit:
Renamed local image name for universal forwarders to be consistent with
make commands and less verbose:

* splunkforwarder-debian-9 => uf-debian-9
* splunkforwarder-centos-7 => uf-centos-7

Added a "bare" build stage to universal forwarder debian-9: ufbare-debian-9

Moved updateetc.sh script up layers from bare-debian-9 to minimal-debian-9

Moved checkstate.sh script down layers from bare-debian-9 to splunk-debian-9

bb03 commented:
@mikedickey Can you address the CI issues to make the build green before we submit?

Commit:
Removed docker collection TAs from UF

Commit:
Switched from using sha512 to md5 for download checksums, since this seems
to be the only method that works reliably across Splunk artifact systems

Removed SPLUNK_FILENAME as a make and docker build argument since it is
no longer necessary, and just makes it harder to use alternate URLs

Fixed permissions of sbin scripts which were causing test failures

s/splunkforwarder/uf/ in test scripts to accomodate image name change

Commit:
Merge remote-tracking branch 'origin/develop' into multistage-build

Commit:
Consolidated Linux platform Dockerfiles to use common multi-stage build templates

Commit:
Fixed broken test case on debian 10

Added missing python-yaml package for debian 10

Shrunk debian-10 base by about 20MB by using pip to install ansible instead of
using the dpkg, which brings with it a ton of dependencies including python3

arctan5x commented on splunk/common-files/Dockerfile:
```diff
@@ -0,0 +1,113 @@
+#
+FROM minimal as bare
+COPY --from=package --chown=splunk:splunk /extras /opt
+EXPOSE 4001 8000 8065 8088 8089 8191 9887 9997
```
Per @nwang92's change, can we remove 4001 from this Dockerfile? Nelson's PR: https://github.com/splunk/docker-splunk/pull/162

melugoyal commented on splunk/common-files/Dockerfile:
```diff
@@ -0,0 +1,113 @@
+    SPLUNK_GROUP=splunk \
+    SPLUNK_USER=splunk \
+    TMPSPLUNKDIR=${SPLUNK_HOME}/tmp \
+    TMPETCDIR=${TMPSPLUNKDIR}/etc
```
i think these need to be separated into separate ENV statements. TMPSPLUNKDIR is set to `/tmp` and TMPETCDIR is set to `/etc`
